### ISSUE TITLE
IDT002 on weird inheritance structure

### ISSUE THREAD

**Issue by fakkoweb** [saucecontrol/InheritDoc#11]
Hello,

I have the following class structure (which I simplified):

```
  /// <summary>
  /// Specific type description
  /// </summary>
  public sealed class EntityAddBroadcast : pb::IMessage<EntityAddBroadcast> {
    private static readonly pb::MessageParser<EntityAddBroadcast> _parser = new pb::MessageParser<EntityAddBroadcast>(() => new EntityAddBroadcast());
    private pb::UnknownFieldSet _unknownFields;
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public static pb::MessageParser<EntityAddBroadcast> Parser { get { return _parser; } }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public static pbr::MessageDescriptor Descriptor {
      get { return global::Protobuf.Gen.Descriptor.MessageTypes[16]; }
    }
  }
  
    /// <summary>
    /// Interface for a Protocol Buffers message, supporting
    /// basic operations required for serialization.
    /// </summary>
    public interface IMessage
    {
        /// <summary>
        /// Merges the data from the specified coded input stream with the current message.
        /// </summary>
        /// <remarks>See the user guide for precise merge semantics.</remarks>
        /// <param name="input"></param>
        void MergeFrom(CodedInputStream input);

        /// <summary>
        /// Descriptor for this message. All instances are expected to return the same descriptor,
        /// and for generated types this will be an explicitly-implemented member, returning the
        /// same value as the static property declared on the type.
        /// </summary>
        MessageDescriptor Descriptor { get; }
    }

    /// <summary>
    /// Generic interface for a Protocol Buffers message,
    /// where the type parameter is expected to be the same type as
    /// the implementation class.
    /// </summary>
    /// <typeparam name="T">The message type.</typeparam>
    public interface IMessage<T> : IMessage, IEquatable<T>, IDeepCloneable<T> where T : IMessage<T>
    {
        /// <summary>
        /// Merges the given message into this one.
        /// </summary>
        /// <remarks>See the user guide for precise merge semantics.</remarks>
        /// <param name="message">The message to merge with this one. Must not be null.</param>
        void MergeFrom(T message);
    }
```

I get on the `Description` property of `EntityAddBroadcast` a:

```
[IDT002] No matching documentation could be found for: P:Protobuf.Gen.EntityAddBroadcast.pb::IMessage#Descriptor, which attempts to inherit from: P:IMessage.Descriptor
```

(I have removed the namespaces from the error because I think they are not relevant).

What is causing this "not found" problem? I did not put an explicit <inheritdoc> there, but it still tries to inherit. Maybe it is the layer of `IMessage<T> : IMessage` that is making it fail?

Any solutions that do not involve changing the protobuf generated files?

**Comment by saucecontrol** [comment]
Howdy!

I assume this is a case of a Roslyn Source Generator inserting `<inheritdoc />` willy nilly on generated types/members, whether the parent has docs or not.  That warning is meant to alert you to cases where you explicitly said you wanted to inherit docs, but they aren't on the parent.

You could add the warning to your ignore list, but it would definitely be better to try to distinguish between those two scenarios in the Task if possible.

I think it should be easy enough to identify types that come from Source Generators, based on `[DebuggerNonUserCode]` or some other tagging attribute they use.  I'm not sure how consistent various generators are, so I'll need to gather some samples.  If you have a minimal full repro for this one, that would be helpful.

**Comment by fakkoweb** [comment]
> Howdy!
> 
> I assume this is a case of a Roslyn Source Generator inserting `<inheritdoc />` willy nilly on generated types/members, whether the parent has docs or not. That warning is meant to alert you to cases where you explicitly said you wanted to inherit docs, but they aren't on the parent.
> 
> You could add the warning to your ignore list, but it would definitely be better to try to distinguish between those two scenarios in the Task if possible.
> 
> I think it should be easy enough to identify types that come from Source Generators, based on `[DebuggerNonUserCode]` or some other tagging attribute they use. I'm not sure how consistent various generators are, so I'll need to gather some samples. If you have a minimal full repro for this one, that would be helpful.

Got it. I noticed this only happens when I install the NuGet on a Unity project from Rider IDE. If I install it on a pure .csproj, I do not see the warnings on these types. I suspect maybe Unity includes things I don't want.

Also, reason why I noticed this is because I am making XML-related warnings to become errors, through `<WarningsAsErrors>`. Basically I am imposing the developer to not screw up documentation. In some instances, IDT002 is useful (in the case you described). But in others, it is unintended, like this case.

I don't have a fully repro project as of now, but i might be able to prepare it in the future. Maybe do you have an example on how to filter by attribute and where to place it in a .csproj? (not much experience on .csproj)

**Comment by saucecontrol** [comment]
Thanks for the extra detail.  It seems clear enough this is an issue with generated code vs. user code, so I think your snippet above will do for testing purposes.

> Maybe do you have an example on how to filter by attribute and where to place it in a .csproj? (not much experience on .csproj)

Sorry for the confusion.  I meant that would be something I could add to the library for the Task.  It's not possible with the current version.

I had a look around at various code generators, and there unfortunately doesn't seem to be any consistent way generated types/members are marked.  I saw various combinations of `[Obsolete]`, `[EditorBrowsable(EditorBrowsableState.Never)]`, `[CompilerGenerated]`, `[GeneratedCode]`, and `[DebuggerNonUserCode]`, but there are probably others.  I think it might make sense to just make a configurable whitelist of attributes for ignored types/members for enforcement of the IDTxxxx warnings to future-proof it.  I'll look a bit more and finalize a default list to start with for this feature.

**Comment by fakkoweb** [comment]
Thanks!
I was wondering if you know a way to exclude specific files (or a filter) in a csproj from XML documentation generation. I tried something like:

```
<ItemGroup>
    <Compile Include="*.cs" />
    <Compile Include=".\modules\**\*.cs" >
        <Link />
    </Compile>
    <!-- Exclude files from documentation -->
    <None Update=".\modules\**\*.cs">
        <GenerateDocumentationFile>false</GenerateDocumentationFile>
    </None>
</ItemGroup>
```

But it does not really listen to me, and still triggers warnings in the files I would like to exclude (like the one I put on this issue). Any idea?

**Comment by saucecontrol** [comment]
Yeah, that won't work either, because the docs are generated by the C# compiler during build (using the [/doc](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/output) switch), and all files are passed to the compiler as a single unit, because it produces one .dll (and one .xml) from all of them.

If you want to get behind msbuild curtain a bit to see how things work, the [msbuild structured log viewer](https://msbuildlog.com/) is excellent for that :)

**Comment by saucecontrol** [comment]
Would you be able to try your build with the latest package version from the [CI feed](https://dev.azure.com/saucecontrol/InheritDoc/_artifacts/feed/inheritdoc_ci)?  I've added the feature to ignore warnings for generated code (with a static attribute list for now), and I'd like to make sure it catches the case you found.

Looking at the warning message again, I'm not so sure about the `pb::` in there since that should be a namespace alias and shouldn't show up anywhere in the compiled assembly or docs, unless there's some IL post-processing as part of your build 🤷‍♂️

**Comment by fakkoweb** [comment]
> Would you be able to try your build with the latest package version from the [CI feed](https://dev.azure.com/saucecontrol/InheritDoc/_artifacts/feed/inheritdoc_ci)? I've added the feature to ignore warnings for generated code (with a static attribute list for now), and I'd like to make sure it catches the case you found.
> 
> Looking at the warning message again, I'm not so sure about the `pb::` in there since that should be a namespace alias and shouldn't show up anywhere in the compiled assembly or docs, unless there's some IL post-processing as part of your build 🤷‍♂️

Missed this. Yes I can try with the setup I had before to see what happens.
To use that version I have to specifically select from package versions?

**Comment by saucecontrol** [comment]
Yep, thanks.  You just need to configure the Azure feed and then add a reference to the latest CI package version.  Something like this should do it, unless there's something about Unity projects that makes package restore different (in which case, you can add the CI feed to your `NuGet.config`)

```xml
<PropertyGroup>
  <RestoreAdditionalProjectSources>https://pkgs.dev.azure.com/saucecontrol/InheritDoc/_packaging/inheritdoc_ci/nuget/v3/index.json</RestoreAdditionalProjectSources>
</PropertyGroup>

<ItemGroup>
  <PackageReference Include="SauceControl.InheritDoc" Version="1.4.0-ci*" />
</ItemGroup>
```

**Comment by saucecontrol** [comment]
Closing this since I believe it's resolved in the 1.4.0 package.  If that doesn't take care of you, feel free to re-open.
